compiler/pash_runtime.sh

```diff
@@ -140,38 +145,60 @@ else
     ## Count the execution time
     pash_exec_time_start=$(date +"%s%N")
 
+    
     ## If the compiler failed or if we dry_run the compiler, we have to run the sequential
     if [ "$pash_runtime_return_code" -ne 0 ] || [ "$pash_dry_run_compiler_flag" -eq 1 ]; then
         pash_script_to_execute="${pash_sequential_script_file}"
     else
         pash_script_to_execute="${pash_compiled_script_file}"
     fi
-
-    ##
-    ## (4)
-    ##
-    source "$RUNTIME_DIR/pash_wrap_vars.sh" ${pash_script_to_execute}
+    pash_runtime_final_status=$?
+    
+    # ##
+    # ## (4)
+    # ##
+
+    function clean_up () {
+        echo "Exit:${process_id}" > "$RUNTIME_IN_FIFO"
+    } 
+
+    function run_parallel() {
+        trap clean_up SIGTERM SIGINT EXIT
+        source "$RUNTIME_DIR/pash_wrap_vars.sh" ${pash_script_to_execute} ${process_id}
+        internal_exec_status=$?
+        clean_up
+        (exit $internal_exec_status)
+    }
+
+    # Don't fork if compilation faild. The script might have effects on the shell state.
+    if [ "$pash_runtime_return_code" -ne 0 ] || [ "$pash_parallel_pipelines" -eq 0 ]; then
+        clean_up # Early clean up in case the script effects shell like "break" or "exec"
+        source "$RUNTIME_DIR/pash_wrap_vars.sh" ${pash_script_to_execute} ${process_id}
+    else 
+        # What happens in case of error?
+        run_parallel <&0 &
+    fi
     pash_runtime_final_status=$?
```

Comment:
What would be the correct way to do this? My guess it's just assuming success in this case.

Reply:
We could try to think of an elaborate solution that somehow manages to pipeline the next compilation (to allow for the extra parallelism), but also manages to return the exit status correctly here. This seems hard but also really important.

For example, is there a way to revert the `&` use. Instead of running `run_parallel` using `&`, maybe we can somehow run the next compilation in parallel. This would require complex use of `wait`s and `&` or maybe a complete redesign of the architecture, but it might be necessary (if this here leads to issues).

At the very least, let's add a TODO for now, and then let's think about this in another PR!

Reply:
added a TODO